abci/cmd/api/main_test.go

```diff
@@ -0,0 +1,378 @@
+		uri = "/v1/register"
+		sig = "0x6d8c7bb3292cab67f4814f9c2d1986430bd188b4eadf82a3fdf1e6be10f7599751985388c2a79429ee60761169e4c67e3b453daf88b637d77f87d7be68196b2c1b"
+		res, code = request(router, "POST", uri, map[string]interface{}{
+			"addr": fixture.Bob.Address.Hex(),
+			"sig":  sig,
+		})
+		So(res["error"], ShouldBeNil)
+		So(code, ShouldEqual, http.StatusOK)
+		So(res, ShouldContainKey, "id")
+		appHeight += 2
+		bobID := res["id"].(string)
+
+		if err := rpcclient.WaitForHeight(client, appHeight, nil); err != nil {
+			t.Error(err)
+		}
+
+		uri = "/v1/account_info?identity=" + fixture.Bob.Address.Hex()
+		res, _ = request(router, "GET", uri, nil)
+		So(res["id"], ShouldEqual, bobID)
+		So(res["balance"], ShouldEqual, "200")
+
+		//
+		// Test GET /block
+		//
+		uri = fmt.Sprintf("/v1/block?height=%d", appHeight)
+		res, code = request(router, "GET", uri, nil)
+		So(res["result"], ShouldNotBeNil)
+		So(code, ShouldEqual, http.StatusOK)
+
+		// Missing params
+		uri = "/v1/block"
+		res, code = request(router, "GET", uri, nil)
+		So(res["error"], ShouldNotBeNil)
+		So(code, ShouldEqual, http.StatusBadRequest)
+
+		// Invalid height
+		uri = "/v1/block?height=-1"
+		res, code = request(router, "GET", uri, nil)
+		So(res["error"], ShouldNotBeNil)
+		So(code, ShouldEqual, http.StatusBadRequest)
+
+		uri = "/v1/block?height=999"
+		res, code = request(router, "GET", uri, nil)
+		So(res["error"], ShouldNotBeNil)
+		So(code, ShouldEqual, http.StatusBadRequest)
+
+		//
+		// Test POST /transfer
+		//
+		uri = "/v1/transfer"
+
+		// Missing params
+		res, code = request(router, "POST", uri, map[string]interface{}{})
+		So(res["error"], ShouldNotBeNil)
+		So(code, ShouldEqual, http.StatusBadRequest)
+
+		sig = "0x343db6effdf722054ff57bcdad4d21b7025407557c631e7e4b1cd77411fa4c155e91c2a8fff992792d885c34411dca84d8d2eaa863ff7eecda1f3322be024d071b"
+		params = map[string]interface{}{
+			"fee":      "0",
+			"identity": fixture.Alice.Address.Hex(),
+			"nonce":    1,
+			"to": []map[string]interface{}{
+				{
+					"identity": fixture.Bob.Address.Hex(),
+					"value":    "1",
+				},
+			},
+			"sig": sig,
+		}
+		res, code = request(router, "POST", uri, params)
+		So(res["error"], ShouldBeNil)
+		So(code, ShouldEqual, http.StatusOK)
+		appHeight += 2
+
+		if err := rpcclient.WaitForHeight(client, appHeight, nil); err != nil {
+			t.Error(err)
+		}
+
+		// Duplicated transfer
+		res, code = request(router, "POST", uri, params)
+		So(code, ShouldEqual, http.StatusConflict)
+		So(res["error"], ShouldNotBeNil)
+
+		// Invalid logic
+		params["to"] = []map[string]interface{}{
+			{
+				"identity": fixture.Bob.Address.Hex(),
+				"value":    "999",
+			},
+		}
+		params["sig"] = "0xe65fc167583c2daab404ce80054bfc731d6653d5ee794f8e389a363bcc6b51224827e016c2412c14c36ea4df5dedb06a69d3093714bf6de8519486d3cac472a91b"
+		res, code = request(router, "POST", uri, params)
+		So(res["error"], ShouldNotBeNil)
```

Comment:
Better parse `res["error"]` to confirm the reason for bad request is insufficient balance (e.g. check `Code`)